From: =?utf-8?q?Jonas_Dre=C3=9Fler?= <verdre@v0yd.nl>
Date: Sat, 19 Dec 2020 17:08:27 +0100
Subject: clutter: Use G_PARAM_EXPLICIT_NOTIFY for actor and stage properties

Turns out GObject implicitly notifies all properties by default as soon
as a property setter is called, no matter if the property actually
changed or not. One can opt-out of this behavior by setting the
G_PARAM_EXPLICIT_NOTIFY flag.

So since almost all our properties get notified explicitely (well,
except ClutterActors deprecated show-on-set-parent property), set this
flag for all properties of ClutterActor and ClutterStage now. This
significantly reduces the number of notify:: signals emitted on
ClutterActors, because in gnome-shell javascript we usually set GObject
properties directly, not by going through the extra setter method.

More cleanups can be done in the future, since this flag is suitable for
almost every property in Clutter and even Mutter.

This fixes a crash where we'd hit a newly introduced assertion in
on_device_actor_reactive_changed() of ClutterStage because
notify::reactive got emitted multiple times.

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1651>
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/mutter/+bug/1918033
---
 clutter/clutter/clutter-actor.c | 147 +++++++++++++++++++++++++++++-----------
 clutter/clutter/clutter-stage.c |   9 ++-
 2 files changed, 113 insertions(+), 43 deletions(-)

diff --git a/clutter/clutter/clutter-actor.c b/clutter/clutter/clutter-actor.c
index ff57017..0c09d5a 100644
--- a/clutter/clutter/clutter-actor.c
+++ b/clutter/clutter/clutter-actor.c
@@ -6072,6 +6072,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.0,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6091,6 +6092,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.0,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6113,6 +6115,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         GRAPHENE_TYPE_POINT,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6132,6 +6135,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.0,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6151,6 +6155,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.0,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6172,6 +6177,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         GRAPHENE_TYPE_SIZE,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6190,7 +6196,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         P_("Forced X position of the actor"),
                         -G_MAXFLOAT, G_MAXFLOAT,
                         0.0,
-                        CLUTTER_PARAM_READWRITE);
+                        CLUTTER_PARAM_READWRITE |
+                        G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:fixed-y:
@@ -6208,7 +6215,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         P_("Forced Y position of the actor"),
                         -G_MAXFLOAT, G_MAXFLOAT,
                         0,
-                        CLUTTER_PARAM_READWRITE);
+                        CLUTTER_PARAM_READWRITE |
+                        G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:fixed-position-set:
@@ -6223,7 +6231,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Fixed position set"),
                           P_("Whether to use fixed positioning for the actor"),
                           FALSE,
-                          CLUTTER_PARAM_READWRITE);
+                          CLUTTER_PARAM_READWRITE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:min-width:
@@ -6243,7 +6252,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         P_("Forced minimum width request for the actor"),
                         0.0, G_MAXFLOAT,
                         0.0,
-                        CLUTTER_PARAM_READWRITE);
+                        CLUTTER_PARAM_READWRITE |
+                        G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:min-height:
@@ -6262,7 +6272,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         P_("Forced minimum height request for the actor"),
                         0.0, G_MAXFLOAT,
                         0.0,
-                        CLUTTER_PARAM_READWRITE);
+                        CLUTTER_PARAM_READWRITE |
+                        G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:natural-width:
@@ -6281,7 +6292,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         P_("Forced natural width request for the actor"),
                         0.0, G_MAXFLOAT,
                         0.0,
-                        CLUTTER_PARAM_READWRITE);
+                        CLUTTER_PARAM_READWRITE |
+                        G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:natural-height:
@@ -6300,7 +6312,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         P_("Forced natural height request for the actor"),
                         0.0, G_MAXFLOAT,
                         0.0,
-                        CLUTTER_PARAM_READWRITE);
+                        CLUTTER_PARAM_READWRITE |
+                        G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:min-width-set:
@@ -6315,7 +6328,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Minimum width set"),
                           P_("Whether to use the min-width property"),
                           FALSE,
-                          CLUTTER_PARAM_READWRITE);
+                          CLUTTER_PARAM_READWRITE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:min-height-set:
@@ -6330,7 +6344,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Minimum height set"),
                           P_("Whether to use the min-height property"),
                           FALSE,
-                          CLUTTER_PARAM_READWRITE);
+                          CLUTTER_PARAM_READWRITE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:natural-width-set:
@@ -6345,7 +6360,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Natural width set"),
                           P_("Whether to use the natural-width property"),
                           FALSE,
-                          CLUTTER_PARAM_READWRITE);
+                          CLUTTER_PARAM_READWRITE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:natural-height-set:
@@ -6360,7 +6376,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Natural height set"),
                           P_("Whether to use the natural-height property"),
                           FALSE,
-                          CLUTTER_PARAM_READWRITE);
+                          CLUTTER_PARAM_READWRITE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:allocation:
@@ -6379,6 +6396,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         CLUTTER_TYPE_ACTOR_BOX,
                         G_PARAM_READABLE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6446,7 +6464,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                        P_("The actor's request mode"),
                        CLUTTER_TYPE_REQUEST_MODE,
                        CLUTTER_REQUEST_HEIGHT_FOR_WIDTH,
-                       CLUTTER_PARAM_READWRITE);
+                       CLUTTER_PARAM_READWRITE |
+                       G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:z-position:
@@ -6473,6 +6492,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.0f,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6491,6 +6511,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                        255,
                        G_PARAM_READWRITE |
                        G_PARAM_STATIC_STRINGS |
+                       G_PARAM_EXPLICIT_NOTIFY |
                        CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6524,7 +6545,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Visible"),
                           P_("Whether the actor is visible or not"),
                           FALSE,
-                          CLUTTER_PARAM_READWRITE);
+                          CLUTTER_PARAM_READWRITE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:mapped:
@@ -6539,7 +6561,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Mapped"),
                           P_("Whether the actor will be painted"),
                           FALSE,
-                          CLUTTER_PARAM_READABLE);
+                          CLUTTER_PARAM_READABLE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:realized:
@@ -6553,7 +6576,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Realized"),
                           P_("Whether the actor has been realized"),
                           FALSE,
-                          CLUTTER_PARAM_READABLE);
+                          CLUTTER_PARAM_READABLE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:reactive:
@@ -6569,7 +6593,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Reactive"),
                           P_("Whether the actor is reactive to events"),
                           FALSE,
-                          CLUTTER_PARAM_READWRITE);
+                          CLUTTER_PARAM_READWRITE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:has-clip:
@@ -6581,7 +6606,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Has Clip"),
                           P_("Whether the actor has a clip set"),
                           FALSE,
-                          CLUTTER_PARAM_READABLE);
+                          CLUTTER_PARAM_READABLE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:clip-rect:
@@ -6602,7 +6628,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         P_("The visible region of the actor"),
                         GRAPHENE_TYPE_RECT,
                         G_PARAM_READWRITE |
-                        G_PARAM_STATIC_STRINGS);
+                        G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:name:
@@ -6616,7 +6643,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          P_("Name"),
                          P_("Name of the actor"),
                          NULL,
-                         CLUTTER_PARAM_READWRITE);
+                         CLUTTER_PARAM_READWRITE |
+                         G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:pivot-point:
@@ -6640,6 +6668,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         GRAPHENE_TYPE_POINT,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6660,6 +6689,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.f,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6679,6 +6709,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          1.0,
                          G_PARAM_READWRITE |
                          G_PARAM_STATIC_STRINGS |
+                         G_PARAM_EXPLICIT_NOTIFY |
                          CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6698,6 +6729,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          1.0,
                          G_PARAM_READWRITE |
                          G_PARAM_STATIC_STRINGS |
+                         G_PARAM_EXPLICIT_NOTIFY |
                          CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6717,6 +6749,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          1.0,
                          G_PARAM_READWRITE |
                          G_PARAM_STATIC_STRINGS |
+                         G_PARAM_EXPLICIT_NOTIFY |
                          CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6736,6 +6769,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          0.0,
                          G_PARAM_READWRITE |
                          G_PARAM_STATIC_STRINGS |
+                         G_PARAM_EXPLICIT_NOTIFY |
                          CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6755,6 +6789,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          0.0,
                          G_PARAM_READWRITE |
                          G_PARAM_STATIC_STRINGS |
+                         G_PARAM_EXPLICIT_NOTIFY |
                          CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6794,6 +6829,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.f,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6814,6 +6850,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.f,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6834,6 +6871,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.f,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6864,6 +6902,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         CLUTTER_TYPE_MATRIX,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6879,7 +6918,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Whether the transform property is set"),
                           FALSE,
                           G_PARAM_READABLE |
-                          G_PARAM_STATIC_STRINGS);
+                          G_PARAM_STATIC_STRINGS |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:child-transform:
@@ -6902,6 +6942,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         CLUTTER_TYPE_MATRIX,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -6917,7 +6958,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Whether the child-transform property is set"),
                           FALSE,
                           G_PARAM_READABLE |
-                          G_PARAM_STATIC_STRINGS);
+                          G_PARAM_STATIC_STRINGS |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:show-on-set-parent:
@@ -6952,7 +6994,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Clip to Allocation"),
                           P_("Sets the clip region to track the actor's allocation"),
                           FALSE,
-                          CLUTTER_PARAM_READWRITE);
+                          CLUTTER_PARAM_READWRITE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:text-direction:
@@ -6967,7 +7010,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                        P_("Direction of the text"),
                        CLUTTER_TYPE_TEXT_DIRECTION,
                        CLUTTER_TEXT_DIRECTION_LTR,
-                       CLUTTER_PARAM_READWRITE);
+                       CLUTTER_PARAM_READWRITE |
+                       G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:has-pointer:
@@ -6982,7 +7026,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Has Pointer"),
                           P_("Whether the actor contains the pointer of an input device"),
                           FALSE,
-                          CLUTTER_PARAM_READABLE);
+                          CLUTTER_PARAM_READABLE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:actions:
@@ -6996,7 +7041,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          P_("Actions"),
                          P_("Adds an action to the actor"),
                          CLUTTER_TYPE_ACTION,
-                         CLUTTER_PARAM_WRITABLE);
+                         CLUTTER_PARAM_WRITABLE |
+                         G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:constraints:
@@ -7010,7 +7056,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          P_("Constraints"),
                          P_("Adds a constraint to the actor"),
                          CLUTTER_TYPE_CONSTRAINT,
-                         CLUTTER_PARAM_WRITABLE);
+                         CLUTTER_PARAM_WRITABLE |
+                         G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:effect:
@@ -7024,7 +7071,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          P_("Effect"),
                          P_("Add an effect to be applied on the actor"),
                          CLUTTER_TYPE_EFFECT,
-                         CLUTTER_PARAM_WRITABLE);
+                         CLUTTER_PARAM_WRITABLE |
+                         G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:layout-manager:
@@ -7039,7 +7087,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          P_("Layout Manager"),
                          P_("The object controlling the layout of an actor's children"),
                          CLUTTER_TYPE_LAYOUT_MANAGER,
-                         CLUTTER_PARAM_READWRITE);
+                         CLUTTER_PARAM_READWRITE |
+                         G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:x-expand:
@@ -7055,7 +7104,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Whether extra horizontal space should be assigned to the actor"),
                           FALSE,
                           G_PARAM_READWRITE |
-                          G_PARAM_STATIC_STRINGS);
+                          G_PARAM_STATIC_STRINGS |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:y-expand:
@@ -7071,7 +7121,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Whether extra vertical space should be assigned to the actor"),
                           FALSE,
                           G_PARAM_READWRITE |
-                          G_PARAM_STATIC_STRINGS);
+                          G_PARAM_STATIC_STRINGS |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:x-align:
@@ -7088,7 +7139,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                        P_("The alignment of the actor on the X axis within its allocation"),
                        CLUTTER_TYPE_ACTOR_ALIGN,
                        CLUTTER_ACTOR_ALIGN_FILL,
-                       CLUTTER_PARAM_READWRITE);
+                       CLUTTER_PARAM_READWRITE |
+                       G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:y-align:
@@ -7104,7 +7156,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                        P_("The alignment of the actor on the Y axis within its allocation"),
                        CLUTTER_TYPE_ACTOR_ALIGN,
                        CLUTTER_ACTOR_ALIGN_FILL,
-                       CLUTTER_PARAM_READWRITE);
+                       CLUTTER_PARAM_READWRITE |
+                       G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:margin-top:
@@ -7126,6 +7179,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.0,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -7148,6 +7202,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.0,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -7170,6 +7225,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.0,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -7192,6 +7248,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         0.0,
                         G_PARAM_READWRITE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -7206,7 +7263,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                           P_("Background Color Set"),
                           P_("Whether the background color is set"),
                           FALSE,
-                          CLUTTER_PARAM_READABLE);
+                          CLUTTER_PARAM_READABLE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:background-color:
@@ -7225,6 +7283,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                               CLUTTER_COLOR_Transparent,
                               G_PARAM_READWRITE |
                               G_PARAM_STATIC_STRINGS |
+                              G_PARAM_EXPLICIT_NOTIFY |
                               CLUTTER_PARAM_ANIMATABLE);
 
   /**
@@ -7239,7 +7298,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          P_("First Child"),
                          P_("The actor's first child"),
                          CLUTTER_TYPE_ACTOR,
-                         CLUTTER_PARAM_READABLE);
+                         CLUTTER_PARAM_READABLE |
+                         G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:last-child:
@@ -7253,7 +7313,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          P_("Last Child"),
                          P_("The actor's last child"),
                          CLUTTER_TYPE_ACTOR,
-                         CLUTTER_PARAM_READABLE);
+                         CLUTTER_PARAM_READABLE |
+                         G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:content:
@@ -7268,7 +7329,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                          P_("Content"),
                          P_("Delegate object for painting the actor's content"),
                          CLUTTER_TYPE_CONTENT,
-                         CLUTTER_PARAM_READWRITE);
+                         CLUTTER_PARAM_READWRITE |
+                         G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:content-gravity:
@@ -7295,7 +7357,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                        P_("Alignment of the actor's content"),
                        CLUTTER_TYPE_CONTENT_GRAVITY,
                        CLUTTER_CONTENT_GRAVITY_RESIZE_FILL,
-                       CLUTTER_PARAM_READWRITE);
+                       CLUTTER_PARAM_READWRITE |
+                       G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:content-box:
@@ -7317,6 +7380,7 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         CLUTTER_TYPE_ACTOR_BOX,
                         G_PARAM_READABLE |
                         G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY |
                         CLUTTER_PARAM_ANIMATABLE);
 
   obj_props[PROP_MINIFICATION_FILTER] =
@@ -7325,7 +7389,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                        P_("The filter used when reducing the size of the content"),
                        CLUTTER_TYPE_SCALING_FILTER,
                        CLUTTER_SCALING_FILTER_LINEAR,
-                       CLUTTER_PARAM_READWRITE);
+                       CLUTTER_PARAM_READWRITE |
+                       G_PARAM_EXPLICIT_NOTIFY);
 
   obj_props[PROP_MAGNIFICATION_FILTER] =
     g_param_spec_enum ("magnification-filter",
@@ -7333,7 +7398,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                        P_("The filter used when increasing the size of the content"),
                        CLUTTER_TYPE_SCALING_FILTER,
                        CLUTTER_SCALING_FILTER_LINEAR,
-                       CLUTTER_PARAM_READWRITE);
+                       CLUTTER_PARAM_READWRITE |
+                       G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterActor:content-repeat:
@@ -7349,7 +7415,8 @@ clutter_actor_class_init (ClutterActorClass *klass)
                         CLUTTER_TYPE_CONTENT_REPEAT,
                         CLUTTER_REPEAT_NONE,
                         G_PARAM_READWRITE |
-                        G_PARAM_STATIC_STRINGS);
+                        G_PARAM_STATIC_STRINGS |
+                        G_PARAM_EXPLICIT_NOTIFY);
 
   g_object_class_install_properties (object_class, PROP_LAST, obj_props);
 
diff --git a/clutter/clutter/clutter-stage.c b/clutter/clutter/clutter-stage.c
index 0444c0d..b825396 100644
--- a/clutter/clutter/clutter-stage.c
+++ b/clutter/clutter/clutter-stage.c
@@ -1786,7 +1786,8 @@ clutter_stage_class_init (ClutterStageClass *klass)
                           P_("Perspective"),
                           P_("Perspective projection parameters"),
                           CLUTTER_TYPE_PERSPECTIVE,
-                          CLUTTER_PARAM_READABLE);
+                          CLUTTER_PARAM_READABLE |
+                          G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterStage:title:
@@ -1800,7 +1801,8 @@ clutter_stage_class_init (ClutterStageClass *klass)
                            P_("Title"),
                            P_("Stage Title"),
                            NULL,
-                           CLUTTER_PARAM_READWRITE);
+                           CLUTTER_PARAM_READWRITE |
+                           G_PARAM_EXPLICIT_NOTIFY);
 
   /**
    * ClutterStage:key-focus:
@@ -1817,7 +1819,8 @@ clutter_stage_class_init (ClutterStageClass *klass)
                            P_("Key Focus"),
                            P_("The currently key focused actor"),
                            CLUTTER_TYPE_ACTOR,
-                           CLUTTER_PARAM_READWRITE);
+                           CLUTTER_PARAM_READWRITE |
+                           G_PARAM_EXPLICIT_NOTIFY);
 
   g_object_class_install_properties (gobject_class, PROP_LAST, obj_props);
 
